File: backend/repositories/User.repo.go
package repositories

import (
	"github.com/jmoiron/sqlx"
	"github.com/lxndrrud/blog-project/models"
)

type IUserRepo interface {
	GetUserById(idUser int64) (models.User, error)
	GetUserByLogin(login string) (models.User, error)
	InsertUser(login, password string) (int64, error)
}

func NewUserRepo(db *sqlx.DB) IUserRepo {
	return &userRepo{
		db: db,
	}
}

type userRepo struct {
	db *sqlx.DB
}

func (c *userRepo) GetUserById(idUser int64) (models.User, error) {
	var user models.User

	err := c.db.Get(
		&user,
		`
			SELECT 
				u.id as user_id, u.login as user_login, u.password as user_password
			FROM "public"."users" u
			WHERE u.id = $1
		`,
		idUser)
	return user, err
}

func (c *userRepo) GetUserByLogin(login string) (models.User, error) {
	var user models.User

	err := c.db.Get(
		&user,
		`
			SELECT 
				u.id as user_id, u.login as user_login, u.password as user_password
			FROM public.users u
			WHERE login = $1
		`,
		login,
	)

	return user, err
}

func (c *userRepo) InsertUser(login, password string) (int64, error) {
	var idUser int64
	err := c.db.Get(&idUser,
		`
			INSERT INTO public.users (login, password, id_role) VALUES
			($1, $2, 2)
			RETURNING id;
		`,
		login, password)
	return idUser, err
}

File: backend/src/repositories/User.repo.ts
import { Knex } from "knex";

export interface IUserRepo {
    getUserById(idUser: number): Promise<{
        id: number;
        login: string;
        password: string;
    } | undefined>
    getUserByLogin(login: string): Promise<{
        id: number;
        login: string;
        password: string;
    } | undefined>
    insertUser(payload: {
        login: string;
        password: string;
    }): Promise<number[]>
}

export class UserPostgresRepo implements IUserRepo {
    constructor(
        private readonly connection: Knex
    ) {}

    public async getUserById(idUser: number) {
        const user: {
            id: number,
            login: string,
            password: string
        } | undefined = await this.connection('public.users as u')
            .where('u.id', idUser)
            .first()
        return user
    }

    public async getUserByLogin(login: string) {
        const user: {
            id: number,
            login: string,
            password: string
        } | undefined = await this.connection('public.users as u')
            .where('u.login', login)
            .first()
        return user
    }

    public async insertUser(payload: { login: string, password: string }) {
        return this.connection('public.users')
            .returning('id')
            .insert({
                ...payload, id_role: 2
            })
    }
}
